### PR TITLE
Improve how some kind of Expr are displayed by the show method

### PR DESCRIPTION
**Pull request by ChiaffarelliMarco**
The improvements concerns the following Expr:
- typed bracket expressions as
`T[a b]
T[a b; c d]`
- quote blocks + unquoting
`quote    
    $a+$b    
end`
-"special macros"
`` `a b c`    
r"foo"    
foo"bar"zot``
-"long" numeric literals, which are implemented as macros
`11111111111111111111    
0xfffffffffffffffff    
11111111111111111111111111111111111111111111111111111111111111`

Also, now the function test_repr in test/show.jl can (and indeed does) directly and successfully compare
`Meta.parse(x)`
with
`eval(Meta.parse(repr(Meta.parse(x))))`
Notice that I'm not to be credited with the latter achievement, since master branch also could have
(but did not) directly and successfully use the same comparison. I suppose the person to be credited
with solving the issue preventing this way of testing to succeed failed to actually update the test routine.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
This is awesome, thanks! Welcome.

File: base/show.jl [hunk 1]
```diff
@@ -1118,6 +1118,15 @@ function show_import_path(io::IO, ex)
     end
 end
 
+function is_core_macro(arg, macro_name::AbstractString)
+    arg === Symbol(macro_name) ||
+    arg === GlobalRef(Core, Symbol(macro_name)) ||
```
Comment (JeffBezanson) [inline base/show.jl]: These are always parsed as `GlobalRef` so I'm not sure the other forms are necessary or desirable.

File: base/show.jl [hunk 2]
```diff
@@ -1350,19 +1374,40 @@ function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int)
...
-            print(io, args[2], ' ')
-        end
-        # Use the functional syntax unless specifically designated with prec=-1
-        # and hide the line number argument from the argument list
-        if prec >= 0
-            show_call(io, :call, args[1], args[3:end], indent)
+        # handle some special syntax
+        # x"y" and x"y"z
+        if occursin(r"^Symbol\(\"@[a-zA-Z].*_str\"\)$",
+                    repr(args[1]))
```
Comment (JeffBezanson) [inline base/show.jl]: I think `endswith(string(args[1]::Symbol), "_str")` should be sufficient; would be better not to depend on regexes too much in this code.

File: test/docs.jl [hunk 3]
```diff
@@ -1157,5 +1157,6 @@ for fn in (:isdone,)
     global xs = @foo $fn
 end
 end
-@test M27832.xs == ":(\$(Expr(:\$, :fn)))"
+# @test M27832.xs == ":(\$(Expr(:\$, :fn)))"
```
Comment (JeffBezanson) [inline test/docs.jl]: Can remove the old test.

File: test/show.jl [hunk 4]
```diff
@@ -58,10 +58,12 @@ function test_repr(x::String)
     # strings get converted to string Exprs by the first show().
     # This could produce a few false positives, but until string
     # interpolation works we don't really have a choice.
+    #
+    # Rectification: comparing x1 and x2 seans to be working
     x1 = Meta.parse(x)
     x2 = eval(Meta.parse(repr(x1)))
     x3 = eval(Meta.parse(repr(x2)))
-    if x3 != x1
+    if x2 != x1
```
Comment (JeffBezanson) [inline test/show.jl]: Test both?

File: base/show.jl (hunk 1 above)
Comment (ChiaffarelliMarco) [reply base/show.jl]: Well, thanks for the almost realtime feedback!

> These are always parsed as `GlobalRef`

Yes, I did guess that, and that was the only case I had initially covered,
but as things stand it is the case that for example the following
`eval(Expr(:macrocall, GlobalRef(Core, Symbol("@int128_str")), LineNumberNode(1, :none), "111111111111111111111"))`
and 
`eval(Expr(:macrocall, Expr(:(.), Core, Expr(:quote, Symbol("@int128_str"))), nothing, "111111111111111111111"))`
compare equal according to `==` (that is both are valid Expr and they evaluate to the same number)  which is why I added the other forms, which cover the latter and other similar cases. Of course this decision may be debatable, and it also occurs to me that the exhibited behavior may be unintended, even though it is consistent with what is declared about macrocalls in the devdocs at [https://pkg.julialang.org/docs/julia/THl1k/1.1.1/devdocs/ast.html#Surface-syntax-AST-1](url)
Let me know which is your stand on this matter.

File: base/show.jl (hunk 2 above)
Comment (ChiaffarelliMarco) [reply base/show.jl]: Ok, modified accordingly.

File: test/show.jl (hunk 4 above)
Comment (ChiaffarelliMarco) [reply test/show.jl]: Sure, that cannot hurt

File: base/show.jl (hunk 1 above)
Comment (JeffBezanson) [reply base/show.jl]: Well, lots of things are equal after calling `eval` :) I think printing those expressions differently makes sense, since one corresponds to how we parse `111111111111111111111` and one does not.

Comment (ChiaffarelliMarco) [reply base/show.jl]: I see your point, updated accordingly.

**Comment by ChiaffarelliMarco** [comment]
#32412

**Comment by ChiaffarelliMarco** [comment]
I see that having merged with #32412 fixed only the buildbot/tester_win32 run, but the buildbot/tester_win64 still fails, even though with different symptoms: this time it doesn't shows any error(but it shows some warnings), it simply time out at some point.
As I cannot test this stuff on a win64 machine, I'm a little unsure about what should I do now.

**Comment by musm** [comment]
That failure simply looks like a fluke: it timed out. We see this from time to time, so I wouldn't worry about it.

**Comment by ChiaffarelliMarco** [comment]
Thanks @musm for the reply. Is there something specific I should do, or the failing test will at some point be subject to a rebuilt? Or could I forget about it and leave it in this failed state?

File: base/show.jl [hunk 5]
```diff
@@ -1350,19 +1369,41 @@ function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int)
...
+        elseif is_core_macro(args[1], "@int128_str") ||
+               is_core_macro(args[1], "@uint128_str") ||
+               is_core_macro(args[1], "@big_str")
+            print(io, args[3])
+        # x"y" and x"y"z
+        elseif isa(args[1], Symbol) &&
+               startswith(string(args[1]::Symbol), "@") &&
+               endswith(string(args[1]::Symbol), "_str")
+            m = match(r"^Symbol\(\"@(.+)_str\"\)$",
+                      repr(args[1]))
```
Comment (JeffBezanson) [inline base/show.jl]: This can use `str[2:prevind(a,end,4)]`.

**Comment by JeffBezanson** [comment]
Could you rebase this now that #32412 is merged?

File: test/show.jl [hunk 6]
```diff
@@ -58,10 +58,12 @@ function test_repr(x::String)
     # strings get converted to string Exprs by the first show().
     # This could produce a few false positives, but until string
     # interpolation works we don't really have a choice.
+    #
+    # Rectification: comparing x1 and x2 seans to be working
```
Comment (c42f) [inline test/show.jl]: Is it possible to remove the comment as a whole then? I must admit I find the original context of this a bit confusing.

**Review by c42f — COMMENTED** [review]
This is a really great first contribution. Thanks!

I'm a little worried about the change to the printing of `$`. Does it make sense for nested quoted code?

**Comment by c42f** [comment]
Just to link up some threads, here's some related (complementary) improvements to show I've been toying with: #32408, https://github.com/JuliaLang/julia/pull/32201#issuecomment-505239962

File: test/show.jl [hunk 7]
```diff
@@ -740,6 +762,15 @@ test_mt(show_f5, "show_f5(A::AbstractArray{T,N}, indices::Vararg{$Int,N})")
 @test_repr "{a b c}"
 @test_repr "{a b; c d}"
 
+# typed vcat and hcat
+@test_repr "T[a]"
+@test_repr "T[a,b]"
+@test_repr "T[a;b;c]"
+@test_repr "T[a b]"
+# @test_repr "T[a b;]" #= don't now what to expect from this =#
```
[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply test/show.jl]: Indeed it was a small bug (now fixed) which prevented the correct displaying of `t[a b;]` as is and without other interventions.

File: base/show.jl [hunk 8]
```diff
@@ -1419,7 +1469,8 @@ function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int)
         end
         print(io, '"')
 
-    elseif (head === :&#= || head === :$=#) && nargs == 1
+    # elseif (head === :&#= || head === :$=#) && nargs == 1
```
[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply base/show.jl]: Me too I'm a little bit confused about this issue. In any case, I've added two test case for nested quotes and interpolations, which pass, but I'm not sure that this matter is completely addressed.

File: test/show.jl [hunk 9]
```diff
@@ -760,6 +791,66 @@ end"""
         \$a + \$b
     end
 end"""
+@test repr(Meta.parse(
+"""macro m(a, b)
+    quote
+        \$a + \$b
+    end
+end""")) ==
+":(macro m(a, b)\n      #= none:2 =#\n      quote\n          #= none:3 =#\n          \$a + \$b\n      end\n  end)"
```
[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply test/show.jl]: Yes, I'm with you. Updated accordingly.

File: test/show.jl (hunk 6 above)
Comment (ChiaffarelliMarco) [reply test/show.jl]: Yes, I think it could be removed, but for now I left it there because I'm somehow confused too about this.
Maybe @JeffBezanson  could provide some thought on this matter? (and also about nested quotes and interpolations)

File: base/show.jl (hunk 5 above)
Comment (ChiaffarelliMarco) [reply base/show.jl]: Updated accordingly (with due changes).

**Comment by ChiaffarelliMarco** [comment]
Rebased as requested to accomodate for #32412 having been merged.

File: base/show.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (c42f) [reply base/show.jl]: I read through that again. Vaguely, I think the problem is that the fallback for printing `Expr`s with unknown head (or other non-surface syntax) is currently to interpolate in an unquoted `Expr` constructor with the arguments to that re-quoted again. Then we run into trouble for `$`s in the re-quoted section. I haven't got an example yet though, and it looks like the expander (`julia-bq-expand` if I'm not mistaken) has changed since then.

[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply base/show.jl]: I've added some test cases to check for possible misbehavior in this scenario, and things seam
to be ok. In the process I've found a problem with the printing of `$x` and the like. More on that issue and its solution in another comment.

**Comment by ChiaffarelliMarco** [comment]
I've realized that printing unquotes as `$x` do not work if they are not properly enclosed in a quote, that is,
for this way of printing to work it is necessary for the number of nested quote to at least match the number of nested unquotes. If this is not the case, it seams to me that the only way forward is to fallback to the "unhandled" case, printing it like `$(Expr(:$, whatever))`.
I've thus added a mechanism that keep track of how many quotes/unquotes has been encountered descending down an Expr tree, triggering the fallback mechanism if it is required, otherwise displaying stuff like `$x` as is.

File: base/show.jl [hunk 10]
```diff
@@ -1347,47 +1371,77 @@ function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int)
...
+        # 11111111111111111111, 0xfffffffffffffffff, 1111...many digits...
+        elseif is_core_macro(args[1], "@int128_str") ||
+               is_core_macro(args[1], "@uint128_str") ||
+               is_core_macro(args[1], "@big_str")
+            print(io, args[3])
+        # x"y" and x"y"z
+        elseif isa(args[1], Symbol) &&
+               startswith(string(args[1]::Symbol), "@") &&
+               endswith(string(args[1]::Symbol), "_str")
+            s = repr(args[1])
```
Comment (c42f) [inline base/show.jl]: `string(::Symbol)` is more efficient in principle than `repr` so you may as well use that if you don't need the `:` prefix. (It looks like it's not as efficient as `String(::Symbol)` right now. Probably for no good reason.)

File: test/show.jl [hunk 11]
```diff
@@ -760,6 +793,131 @@ end"""
...
 end"""
+@test repr(Meta.parse(
+"""macro m(a, b)
+    quote
+        \$a + \$b
+    end
+end""")) ==
+"""
+:(macro m(a, b)
+      #= none:2 =#
```
Comment (c42f) [inline test/show.jl]: If it helps, you can strip line number nodes with `Base.remove_linenums!` so they don't need to be included in the reference string here.

File: base/show.jl [hunk 12]
```diff
@@ -924,36 +925,38 @@ show_linenumber(io::IO, line, file) = print(io, "#= ", file, ":", line, " =#")
...
 end
-show_block(io::IO,head,    block,i::Int) = show_block(io,head, [], block,i)
-function show_block(io::IO, head, arg, block, i::Int)
+show_block(io::IO,head,    block,i::Int, quote_level::Int) = show_block(io,head, [], block,i, quote_level)
+function show_block(io::IO, head, arg, block, i::Int, quote_level::Int)
     if is_expr(arg, :block)
-        show_block(io, head, arg.args, block, i)
+        show_block(io, head, arg.args, block, i, quote_level)
+    elseif is_expr(arg, :quote)
+        show_block(io, head, arg.args, block, i, quote_level+0)
```
Comment (c42f) [inline base/show.jl]: `quote_level+0`?

File: base/download.jl [hunk 13]
```diff
@@ -28,7 +28,7 @@ function find_curl()
         "/usr/bin/curl"
     elseif Sys.iswindows() && Sys.isexecutable(joinpath(get(ENV, "SYSTEMROOT", "C:\\Windows"), "System32\\curl.exe"))
         joinpath(get(ENV, "SYSTEMROOT", "C:\\Windows"), "System32\\curl.exe")
-    elseif Sys.which("curl") !== nothing
+    elseif !Sys.iswindows() && Sys.which("curl") !== nothing
```
Comment (c42f) [inline base/download.jl]: What's this unrelated looking change about?

**Review by c42f — COMMENTED** [review]
It looks to me that you're on the right track to sorting out the problems with `$` using `quote_level`.

Does `quote_level` need to interact with the interpolation syntax used to print unusual `Expr`s in the `unhandled` section and inside `show_unquoted(io::IO, ex::QuoteNode, indent::Int, prec::Int)` ?

File: base/show.jl (hunk 12 above)
Comment (ChiaffarelliMarco) [reply base/show.jl]: Originally it was (wrongly) `quote_level+1`, I've changed it to `quote_level+0` during debug and then i've forgot to commit the change where I've reunited that elseif branch with the previus one.

File: base/download.jl (hunk 13 above)
Comment (ChiaffarelliMarco) [reply base/download.jl]: It's there since the times when #32412 was not merged and I was trying to get CI not to fail, but now that #32412 has been merged I don't know why that change is still shown. In fact on my local repository,
running
`git diff better-show master -- base/download.jl`
gives nothing.

File: test/show.jl (hunk 11 above)
Comment (ChiaffarelliMarco) [reply test/show.jl]: Thanks for the suggestion.

File: base/show.jl (hunk 10 above)
Comment (ChiaffarelliMarco) [reply base/show.jl]: Ok, I've changed it accordingly.

**Comment by ChiaffarelliMarco** [comment]
Regarding @c42f question about the need for interactions between `quote_level` and the `unhandled` and `QuoteNode` case, I think that @c42f is right, as it seams I've missed to pass `quote_level` around in those situations, but preliminary tests show that the 'QuoteNode` case works just fine, contrary to what I had expected. This fact confuse me. I have to look closer, I suspect.

**Comment by c42f** [comment]
`QuoteNode` is for literal symbols and ASTs; it doens't support interpolation.

I mention it because it's one of the expressions which is itself printed as an interpolated expression when it's `show`n and I wondered if that mattered.

```
julia> eval(Expr(:quote, (Expr(:$, :x))))
ERROR: UndefVarError: x not defined
Stacktrace:
 [1] eval at ./boot.jl:328 [inlined]
 [2] eval(::Expr) at ./client.jl:404
 [3] top-level scope at none:0

julia> eval(QuoteNode((Expr(:$, :x))))
:($(Expr(:$, :x)))
```

**Comment by c42f** [comment]
Do you have more time to push this forward? It's such a big improvement it would be a pity to not finish it off. It would be good to rebase to fix the conflicts, and maybe beneficial to squash the commits while you're at it.

After that, there's one remaining question in my mind: does this printing handle all cases where exotic `Expr` heads are mixed in with nested quoting and interpolation? I need to try this branch out, but I'm thinking of things vaguely like

```julia
Expr(:block, Expr(:exotic_head, Expr(:$, :x)))
```

**Comment by ChiaffarelliMarco** [comment]
I apologies for the long absence and for the late reply: in these past moths i was really busy in completing my master degree, and this will continue until the end of October. After that I think I should be able to find some time to reprise this PR.

**Comment by c42f** [comment]
No problem at all!

File: base/show.jl [hunk 14]
```diff
@@ -1114,12 +1118,17 @@ function show_import_path(io::IO, ex)
...
 end
 
+function is_core_macro(arg, macro_name::AbstractString)
+    arg === GlobalRef(Core, Symbol(macro_name))
+end
+
 # TODO: implement interpolated strings
-function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int)
+show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int) = show_unquoted(io, ex, indent, prec, 0)
+function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int, quote_level::Int)
```
Comment (JeffBezanson) [inline base/show.jl]:
```
function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int, quote_level::Int = 0)
```

File: base/show.jl [hunk 15]
```diff
@@ -1150,13 +1159,25 @@ function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int)
...
+            show_list(io, args, head_, indent, func_prec, quote_level, true)
         end
 
     # list (i.e. "(1, 2, 3)" or "[1, 2, 3]")
-    elseif haskey(expr_parens, head)               # :tuple/:vcat
+    elseif haskey(expr_parens, head) ||                          # :tuple/:vcat
+        head === :typed_vcat || head === :typed_hcat
+        # print the type and defer to the untyped case
+        if head === :typed_vcat || head === :typed_hcat
+            print(io, args[1])
```
Comment (JeffBezanson) [inline base/show.jl]: Shouldn't this call `show_unquoted`?

File: base/show.jl [hunk 16]
```diff
@@ -1442,20 +1505,24 @@ function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int)
...
 
-    elseif (head === :import || head === :using) && nargs == 1 &&
-            (valid_import_path(args[1]) ||
-             (Meta.isexpr(args[1], :(:)) && length(args[1].args) > 1 && all(valid_import_path, args[1].args)))
+    elseif (head === :import || head === :using) &&
+           ((nargs == 1 && (valid_import_path(args[1]) ||
+                           (Meta.isexpr(args[1], :(:)) &&
+                            length(args[1].args) > 1 &&
+                            all(valid_import_path, args[1].args)))) ||
+            (all([Meta.isexpr(args[i], :(.)) for i in 1:nargs]) &&
```
Comment (JeffBezanson) [inline base/show.jl]: Doesn't `valid_import_path` already check this?

Comment (ChiaffarelliMarco) [reply base/show.jl]: In fact it does.

**Comment by ChiaffarelliMarco** [comment]
I have added tests regarding the concerns of @c42f about `repr` of code like
`Expr(:block, Expr(:exotic_head, Expr(:$, :x)))`
which highlighted a missing `quote_level+1` in the "unhandled" branc. With this edit things work fine, it seems to me.

Changing subject, when fixing conflicts due to introduction of the var"#funky name" syntax, I assumed that the output of 
`sprint(show, Expr(:macrocall,
                         Expr(:(.), :Core, Expr(:quote, Symbol("@cmd"))),
                         LineNumberNode(0, :none), "a b c"))`
should be 
`":(#= none:0 =# Core.var\"@cmd\" \"a b c\")"`
and not
`":(#= none:0 =# Core.@cmd \"a b c\")"`
but I'm not really sure about this, since I have not really had the time to understand in which contexts `var"@cmd"` should be used instead of `@cmd`. So please let me know if I got it wrong.

**Comment by JeffBezanson** [comment]
`var"@cmd"` parses `@cmd` as if it were a normal identifier, so that would be like writing `Core.x "a b c"`. So it should be `Core.@cmd` to trigger macro call parsing.

**Comment by c42f** [comment]
> which contexts `var"@cmd"` should be used instead of `@cmd`

For more on this, see https://github.com/JuliaLang/julia/blob/f0772c5104435a1a2daec6b4dea96803bb51fe7f/base/docs/basedocs.jl#L990-L1004

**Comment by ChiaffarelliMarco** [comment]
I have handled macrocall expressions whose macroname argument is "qualified", as in
`Expr(:macrocall, Expr(:(.), :A, Expr(:quote, Symbol("@m"))), LineNumberNode(0, :none), :a, :b: )`
or
`Expr(:macrocall, Expr(:(.), :A, Quotenode(Symbol("@m"))), LineNumberNode(0, :none), :a, :b: )`
This fixes for example the `repr` of the aforementioned  `Expr(:macrocall, Expr(:(.), :Core, Expr(:quote, Symbol("@cmd"))), LineNumberNode(0, :none), "a b c")`
to be
`":(#= none:0 =# Core.@cmd \"a b c\")"`

File: base/show.jl [hunk 17]
```diff
@@ -1432,25 +1504,49 @@ function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int)
...
-        show_block(io, "begin", ex, indent)
+        show_block(io, "begin", ex, indent, quote_level)
         print(io, "end")
 
     elseif head === :quote && nargs == 1 && isa(args[1], Symbol)
-        show_unquoted_quote_expr(io, args[1]::Symbol, indent, 0)
+        show_unquoted_quote_expr(io, args[1]::Symbol, indent, 0, quote_level+1)
+    elseif head === :quote && nargs == 1 && Meta.isexpr(args[1], :block)
+        if length(args[1].args) == 1
+            # one argument: Expr(:quote, Expr(:block, a_single_arg))
```
[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply base/show.jl]: On the countrary, I see that this branch is needed to correctly handle things like
`macro m(a, b)
    quote$a + $b
    end
end`

[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply base/show.jl]: On the countrary, tests suggest that this branch is needed to correctly handle cases like
```
macro m(a, b)
    quote
        \$a + \$b
    end
end
```
In fact, without the branch
```
julia> repr(Meta.parse("""
       macro m(a, b)
           quote
               \$a + \$b
           end
       end"""))
":(macro m(a, b)\n      #= none:2 =#\n      :(begin\n            #= none:3 =#\n            \$a + \$b\n        end)\n  end)"
```
while with it in place, more satisfactorily the result is
```
julia> repr(Meta.parse("""
       macro m(a, b)
           quote
               \$a + \$b
           end
       end"""))
":(macro m(a, b)\n      #= none:2 =#\n      quote\n          #= none:3 =#\n          \$a + \$b\n      end\n  end)"
```

Another case that fails without the branch is
```
quote
    quote
        \$\$x
    end
end
```
with both quotes turned into `:(begin ... end)` upon `repr(Meta.parse("..."))`ing it:
```
julia> repr(Meta.parse("""
       quote
           quote
               \$\$x
           end
       end
       """))
":(:(begin\n        #= none:2 =#\n        :(begin\n              #= none:3 =#\n              \$\$x\n          end)\n    end))"
```

[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply base/show.jl]: In fact I've checked now, and actually `"quote \$x end"` is parsed as a `quote Expr` containing a `block Expr`:
```
julia> dump(Meta.parse("quote \$x end"))
Expr
  head: Symbol quote
  args: Array{Any}((1,))
    1: Expr
      head: Symbol block
      args: Array{Any}((2,))
        1: LineNumberNode
          line: Int64 1
          file: Symbol none
        2: Expr
          head: Symbol $
          args: Array{Any}((1,))
            1: Symbol x
```

[the remark this answers is not in the document]
Comment (JeffBezanson) [reply base/show.jl]: Yes, those cases are fine. The problem is when the `block` Expr has one element, e.g. it prints the value of `Expr(:quote, Expr(:block, :(a+b)))` as `:(:(a + b))`.

[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply base/show.jl]: After some thinking, I must say that I very much prefer having
```
julia> show(Expr(:quote, Expr(:block, :(a + b))))
:(:(a + b))
```
than
```
julia> show(Expr(:quote, Expr(:block, :(a + b))))
:(:(begin
        a + b
    end))
```
since the former is coherent with the case in which the block has more than one argument, where, for reasons mentioned in my last two comments, the block itself is actually hidden upon `show`ing:
```
julia> show(Expr(:quote, Expr(:block, :(a + b), :(c + d))))
:(quote
      a + b
      c + d
  end)
```
Nevertheless, the final word is of course yours.

[the remark this answers is not in the document]
Comment (c42f) [reply base/show.jl]: Well, we shouldn't print two ASTs the same way if they're actually different trees. The parser doesn't even emit a block for the syntax `:(a+b)`:

```julia
julia> dump(:(:(a+b)))
Expr
  head: Symbol quote
  args: Array{Any}((1,))
    1: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: Symbol +
        2: Symbol a
        3: Symbol b

julia> dump(Expr(:quote, Expr(:block, :(a+b))))
Expr
  head: Symbol quote
  args: Array{Any}((1,))
    1: Expr
      head: Symbol block
      args: Array{Any}((1,))
        1: Expr
          head: Symbol call
          args: Array{Any}((3,))
            1: Symbol +
            2: Symbol a
            3: Symbol b
```

[the remark this answers is not in the document]
Comment (JeffBezanson) [reply base/show.jl]: Yes, exactly. I don't think we can just drop a `:block` expression; for example people often expect them to be there in order to `push!` statements in. Wouldn't it be best to print it as
```
:(quote
      a + b
  end)
```
? I know there's a difference of whether a line number is present, but I don't think that's as important as keeping the extra block nesting.

[the remark this answers is not in the document]
Comment (c42f) [reply base/show.jl]: Sounds good to me. Another option could be `:(:(a + b;))` which parses as a block with a single entry and no line number node. What I'm not entirely sure about is whether that works in all cases?

[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply base/show.jl]: Now I see your point. After some experimenting I think my preference is for @JeffBezanson 's suggestion. With that, we get a nicely consistent picture:
```
julia> show(Expr(:quote, Expr(:block)))
:(quote
  end)

julia> show(Expr(:quote, Expr(:block, :(a + b))))
:(quote
      a + b
  end)

julia> show(Expr(:quote, Expr(:block, :(a + b), :(c + d))))
:(quote
      a + b
      c + d
  end)
```
@c42f  's suggestion is somewhat cleaner, as there are no line numbers appearing in the call `Meta.parse(repr(Expr(:quote, Expr(:block, :(a + b)))))`
but I find `:(:(a + b;))` and in particular `:(:(a;))` to be a little confusing, and I definitely dislike it if instead of `a + b` or `a` there is some more complicated expression, possibly spread over more than one line: compare
```
julia> show(Expr(:quote, Expr(:block, Expr(:if, :(a == b), :(a + b), :(a - b)))))
:(:(if a == b
        a + b
    else
        a - b
    end;))
```
with
```
julia> show(Expr(:quote, Expr(:block, Expr(:if, :(a == b), :(a + b), :(a - b)))))
:(quote
      if a == b
          a + b
      else
          a - b
      end
  end)
```
By the way, this is my opinion, and I must admit that I may dislike semicolons more than they possibly deserve.

[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply base/show.jl]: I have updated the test_repr macro, adding a stronger form of the previous condition which have to be satisfied for the test to pass. The current solution, which neglect the one-argument inner block, fails the stronger test, while both of the suggestions due to @JeffBezanson and @c42f pass it. Now it is only a matter of deciding which solution to endorse.

[the remark this answers is not in the document]
Comment (ChiaffarelliMarco) [reply base/show.jl]: Once this is settled, with a cleaner local history I will merge with the upstream and fix conflicts.

[the remark this answers is not in the document]
Comment (JeffBezanson) [reply base/show.jl]: Let's use the `quote ... end` syntax as you prefer. I don't think there is any particular promise that `:(x;)` will not include a line number; in fact many people would probably prefer if it did include one.

**Comment by ChiaffarelliMarco** [comment]
I have added to the `test_repr` macro a check regarding "Rule 1", mentioned in a summary comment regarding AST printing, in `base/show.jl` around line 780. This is the rule stating that an `Expr ex` should satisfy `Meta.parse(string(ex)) == ex`.
The test I have added actually check if
`Meta.parse(string(ex)) == ex`
with `ex=Meta.parse(s)` and `s` a `String`, because, in order to keep things simple and avoiding having to add lots of test by hand, I have incorporated the check in the `test_repr` macro, which take a string as an argument.
Furthermore, in a small number of test cases the weaker form
`Base.remove_linenums!(Meta.parse(string(ex))) == ex`
is used, because calling `Meta.parse(string(ex)`  results in line numbers metadata to be added to the AST, thus causing the stronger form to fail.

Having said that, "Rule 1" (or its weaker form) was satisfied for free in all cases bar for `Expr`s outside quotes with `:$` as head: handling this case requires differentiating between `show`/`repr` and `print`/`string`. I provided a simple mechanism that does just that. The way it works is by making show_unquoted never using the fallback "unhandled" case for printing the said `Expr`s when it is invoked by `print` or `string`, while for `show` and `repr` it works as before.
For example we now have
```
julia> repr(Expr(:$, :y))
":(\$(Expr(:\$, :y)))"

julia> string(Expr(:$, :y))
"\$y"
```

**Comment by ChiaffarelliMarco** [comment]
I have cleaned up some stuff, and polished the quote_level mechanism taking advantage of IOContext. I think that now the implementation is much cleaner, as it avoids passing around quote_level everywhere. See if you like it.

With this I think I have reached satisfaction with this PR. Let me know if you think there is more to do, or if you want me to squash some commits.

**Comment by vtjnash** [comment]
It seems strongly unlikely to me that quote_level "belongs" in IOContext. I know it seems like a simpler way to pass around arguments, but it's actually dynamic scope and so has a different use case (although one can usually be used to emulate the other, poorly).

**Comment by c42f** [comment]
@vtjnash I don't disagree, but I'd like to better understand your objection to using `IOContext` here. As far as I can tell this is not so much about dynamic scope, and more about what constitutes the public API? In particular,
* `IOContext` is generally used to provide options to `show` which may be set by the caller when showing a type `T`, and therefore constitute part of the public API of `show(::IO, ::T)`.
* Contrast this with `quote_level`, which is an internal implementation detail of `show(::IO, ::Expr)`.

I suspect variables like `quote_level` and `indent` would be fairly suited to a dynamic scope based implementation if julia provided a natural and efficient implementation of dynamic scope (it doesn't). They just shouldn't be considered part of the public API which makes `IOContext` a bit ill suited to the job. Additionally, attaching them to `IOContext` means they will be inferred as `Any` which is an unfortunate performance pessimization.

Regarding an alternative to using `IOContext`, perhaps `show_unquoted` and related functions could just be enhanced with a `ShowASTContext` or some such; this could be used to also bundle up the `indent` and `prec` arguments which currently need to be passed through in many places? Personally I found the way these are sometimes-but-not-always passed around confusing last time I worked on this code.

**Comment by vtjnash** [comment]
> IOContext is generally used to provide options to show which may be set by the caller when showing a type T, and therefore constitute part of the public API of show(::IO, ::T).

This is subtle, but it's exactly backwards: the IOContext is for setting options when using `IO`. It's the public API of `show(::IO, ::Any)` for when `T` is unknown.

By contrast, that means `quote_level` could realistically be an IOContext parameter, and I've even taken a decent crack at implementing that in https://github.com/JuliaLang/julia/pull/27430 (which needs a pretty solid rebase at this point). But since `show` is usually not supposed to print newlines, it's usually a moot point and sufficient just to pass it around as an argument here.

Yes, they could be paired into a `ShowASTContext` struct, and that's often a nice design. I don't think it'll end up enhancing this code much though, or could even be worse, since the values change so often.

**Comment by c42f** [comment]
> Yes, they could be paired into a `ShowASTContext` struct, and that's often a nice design. I don't think it'll end up enhancing this code much though, or could even be worse, since the values change so often.

Sure, I'd say it might or might not be helpful; a bit hard to judge without trying it out. It might be turn out cleanly if paired with a utility function like `indent(::ShowASTContext)` to increment the indent.

**Comment by JeffBezanson** [comment]
Thanks @ChiaffarelliMarco ! Let's revert the `quote_level` change and then merge this. I don't want to hold back this great work with a debate on that design detail.

**Comment by c42f** [comment]
> I don't want to hold back this great work with a debate on that design detail.

Completely agreed :100:. The benefit of merging this soon is high compared to getting some minor implementation details "just right".

Further cleanup can definitely happen in a separate PR. Doing it separately makes sense anyway if combined with a similar refactor for the other AST-showing context parameters.

**Comment by ChiaffarelliMarco** [comment]
Well, thanks for the feedback. I have reverted to the old quote_level mechanism.
I now understand why `IOContext` is not the correct way of doing things, and I second @c42f  suggestion of introducing a `ShowASTContext` object, paired with some mechanism for cleanly changing it.
Still, the current implementation uses IOContext to distinguish between a call to `show_unquote` coming from `print` and a call coming from `show`. For the moment I think its cleaner this way, rather then having to pass around an another parameter, in particular if soon a `ShowASTContext` will come to light.

**Comment by JeffBezanson** [comment]
Thanks for seeing this through @ChiaffarelliMarco ! It's taken quite a while, but we got there!

**Comment by c42f** [comment]
Yes, great work @ChiaffarelliMarco, I look forward to having these improvements :tada:

**Comment by stevengj** [comment]
Rather than adding a new `quote_level::Int` that needs to get passed around everywhere, it might be nicer to change this (and maybe some of the other context data) to use `IOContext(io, quote_level=>...)` instead.

**Comment by c42f** [comment]
@stevengj this was done previously but removed; see discussion starting at https://github.com/JuliaLang/julia/pull/32423#issuecomment-558392106
